app/src/main/java/nu/yona/app/utils/AppUtils.java

```diff
@@ -626,37 +627,35 @@ private static Intent getVPNIntent(VpnProfile profile, Context context)
 	public static void downloadCertificates()
 	{
 		User user = YonaApplication.getEventChangeManager().getDataState().getUser();
-		if (user != null && user.getLinks() != null)
+		if (user == null || user.getEmbedded() == null || user.getEmbedded().getYonaDevices() == null
+				|| YonaApplication.getEventChangeManager().getSharedPreference().getRootCertPath() != null)
 		{
-			if (user.getLinks().getSslRootCert() != null
-					&& YonaApplication.getEventChangeManager().getSharedPreference().getRootCertPath() == null)
-			{
-				new DownloadFileFromURL(user.getLinks().getSslRootCert().getHref(), new DataLoadListener()
-				{
-					@Override
-					public void onDataLoad(Object result)
-					{
-						if (result != null && !TextUtils.isEmpty(result.toString()))
-						{
-							YonaApplication.getEventChangeManager().getSharedPreference().setRootCertPath(result.toString());
-							YonaApplication.getEventChangeManager().notifyChange(EventChangeManager.EVENT_ROOT_CERTIFICATE_DOWNLOADED, null);
-						}
-						logi(TAG, "Download successful: " + result.toString());
-					}
+			return;
+		}
+		DataLoadListenerImpl dataLoadListener = new DataLoadListenerImpl((result) -> handleDownloadFileFromUrlSuccess(result), (result -> handleDownloadFileFromUrlFailure()), null);
+		new DownloadFileFromURL(user.getSslRootCert(), dataLoadListener);
+	}
 
-					@Override
-					public void onError(Object errorMessage)
-					{
-						loge(TAG, "Download fail");
-						trialCertificateCount++;
-						if (trialCertificateCount < 3)
-						{
-							downloadCertificates();
-						}
-					}
-				});
-			}
+	private static Object handleDownloadFileFromUrlSuccess(Object result)
+	{
+		if (result != null && !TextUtils.isEmpty(result.toString()))
+		{
+			YonaApplication.getEventChangeManager().getSharedPreference().setRootCertPath(result.toString());
+			YonaApplication.getEventChangeManager().notifyChange(EventChangeManager.EVENT_ROOT_CERTIFICATE_DOWNLOADED, null);
+		}
+		logi(TAG, "Download successful: " + result.toString());
+		return null; // Dummy return value, to allow use as data load handler
+	}
+
+	private static Object handleDownloadFileFromUrlFailure()
+	{
+		loge(TAG, "Download fail");
+		trialCertificateCount++;
```

Comment:
Can you also rename this? We don't have trial certificates :)

I suggest naming it ``certificateDownloadAttempts``.